mlflow/java/spark/src/test/scala/org/mlflow/spark/autologging/SparkAutologgingIntegrationSuite.scala

```diff
@@ -0,0 +1,246 @@
+    assert(MlflowSparkAutologgingTestUtils.getListeners(spark).isEmpty)
+    MlflowAutologEventPublisher.stop()
+    assert(MlflowSparkAutologgingTestUtils.getListeners(spark).isEmpty)
+    // Call init() after stop(), verify that we create a new listener
+    MlflowAutologEventPublisher.init()
+    val listeners2 = MlflowSparkAutologgingTestUtils.getListeners(spark)
+    assert(listeners2.length == 1)
+    val listener2 = listeners2.head
+    assert(listener2 != listener1)
+  }
+
+  test("MlflowAutologEventPublisher triggers publishEvent with appropriate arguments " +
+    "when reading datasources corresponding to different formats") {
+      val formatToTestDFs = formatToTablePath.map { case (format, tablePath) =>
+        val baseDf = spark.read.format(format).option("inferSchema", "true")
+          .option("header", "true").load(tablePath)
+        format -> Seq(
+          baseDf,
+          baseDf.filter("number > 0"),
+          baseDf.select("number"),
+          baseDf.limit(2),
+          baseDf.filter("number > 0").select("number").limit(2)
+        )
+      }
+
+        formatToTestDFs.foreach { case (format, dfs) =>
+          dfs.foreach { df =>
+            df.printSchema()
+            MlflowAutologEventPublisher.init()
+            val subscriber = spy(new MockSubscriber())
+            MlflowAutologEventPublisher.register(subscriber)
+            assert(MlflowAutologEventPublisher.subscribers.size == 1)
+            // Read DF
+            df.collect()
+            // Verify events logged
+            Thread.sleep(1000)
+            val tablePath = formatToTablePath(format)
+            val expectedPath = getFileUri(tablePath)
+            verify(subscriber, times(1)).notify(any(), any(), any())
+            verify(subscriber, times(1)).notify(expectedPath, "unknown", format)
+            MlflowAutologEventPublisher.stop()
+          }
+        }
+  }
+
+  test("MlflowAutologEventPublisher triggers publishEvent with appropriate arguments " +
+    "when reading a JOIN of two tables") {
+    val formats = formatToTablePath.keys
+    val leftFormat = formats.head
+    val rightFormat = formats.last
+    val leftPath = formatToTablePath(leftFormat)
+    val rightPath = formatToTablePath(rightFormat)
+    val leftDf = spark.read.format(leftFormat).load(leftPath)
+    val rightDf = spark.read.format(rightFormat).load(rightPath)
+    MlflowAutologEventPublisher.init()
+    val subscriber = spy(new MockSubscriber())
+    MlflowAutologEventPublisher.register(subscriber)
+    leftDf.join(rightDf).collect()
+    // Sleep a second to let the SparkListener trigger read
+    Thread.sleep(1000)
+    verify(subscriber, times(2)).notify(any(), any(), any())
+    verify(subscriber, times(1)).notify(getFileUri(leftPath), "unknown", leftFormat)
+    verify(subscriber, times(1)).notify(getFileUri(rightPath), "unknown", rightFormat)
+  }
+
+  test("MlflowAutologEventPublisher publishes to subscribers in order of registration") {
```

Comment:
is this a requirement?

Reply:
It's not, just added it to help ensure correctness of the [next test](https://github.com/mlflow/mlflow/pull/2181/files#diff-2bc6bc37f0b21089d80596747c27a92cR192) ("MlflowAutologEventPublisher can publish to working subscribers even when others are broken"), which assumes that subscribers are notified in the order they're registered. Can do this more simply by exposing & calling a helper for iterating over subscribers in `MlflowAutologEventPublisher` & mocking it out in the `"MlflowAutologEventPublisher can publish to working subscribers even when others are broken"` test to yield a broken subscriber before a working one, or I could add a comment explaining the purpose of this test / that the behavior it guarantees isn't fundamental to the API (fine with doing either)